src/base/baseBrushTool.js

```diff
@@ -0,0 +1,367 @@
+    context.globalAlpha = configuration.brushAlpha;
+    context.drawImage(this._imageBitmap, canvasTopLeft.x, canvasTopLeft.y, canvasWidth, canvasHeight);
+    context.globalAlpha = 1.0;
+  }
+
+  /**
+   *  Changes the draw color (segmentation) of the tool.
+   *
+   * @protected
+   * @param  {Number} drawId The id of the color (segmentation) to switch to.
+   */
+  _changeDrawColor (drawId) {
+    const configuration = this._configuration;
+    const colormap = external.cornerstone.colors.getColormap(COLOR_MAP_ID);
+
+    configuration.draw = drawId;
+    const colorArray = colormap.getColor(configuration.draw);
+
+    configuration.hoverColor = `rgba(${colorArray[[0]]}, ${colorArray[[1]]}, ${colorArray[[2]]}, 0.8 )`;
+    configuration.dragColor = `rgba(${colorArray[[0]]}, ${colorArray[[1]]}, ${colorArray[[2]]}, 1.0 )`;
+  }
+
+  /**
+  * Event handler for MOUSE_UP during the drawing event loop.
+  *
+  * @protected
+  * @event
+  * @param {Object} evt - The event.
+  */
+  _drawingMouseUpCallback(evt) {
+    const eventData = evt.detail;
+    const element = eventData.element;
+
+    this.configuration.drawing = false;
+    this.configuration.mouseUpRender = true;
+
+    this._stopListeningForMouseUp(element);
+  }
+  /**
+  * Adds modify loop event listeners.
+  *
+  * @protected
+  * @param {Object} element - The viewport element to add event listeners to.
+  * @modifies {element}
+  */
+  _startListeningForMouseUp (element) {
+    element.removeEventListener(EVENTS.MOUSE_UP, this._drawingMouseUpCallback);
+    element.removeEventListener(EVENTS.MOUSE_CLICK, this._drawingMouseUpCallback);
+
+    element.addEventListener(EVENTS.MOUSE_UP, this._drawingMouseUpCallback);
+    element.addEventListener(EVENTS.MOUSE_CLICK, this._drawingMouseUpCallback);
+
+    external.cornerstone.updateImage(element);
+  }
+
+  /**
+  * Adds modify loop event listeners.
+  *
+  * @protected
+  * @param {Object} element - The viewport element to add event listeners to.
+  * @modifies {element}
+  */
+  _stopListeningForMouseUp (element) {
+    element.removeEventListener(EVENTS.MOUSE_UP, this._drawingMouseUpCallback);
+    element.removeEventListener(EVENTS.MOUSE_CLICK, this._drawingMouseUpCallback);
+
+    external.cornerstone.updateImage(element);
+  }
+
+
+  /**
+   * Returns the number of colors in the colormap.
+   *
+   * @static
+   * @protected
+   * @return {Number} The number of colors in the color map.
+   */
+  static _getNumberOfColors () {
+    const colormap = external.cornerstone.colors.getColormap(COLOR_MAP_ID);
+
+    return colormap.getNumberOfColors();
+  }
+
+  /**
+   * Sets the keybinds for the tool. // TODO: If this will be used in other
+   * tools, maybe it should be stored in a different module and imported?
+   *
+   */
+  set keyBinds (keyBinds) {
+    const configuration = this.configuration;
+
+    configuration.keyBinds = keyBinds;
+    this._keyboardController = new KeyboardController(this, keyBinds);
```

Comment:
Oooh. Interesting. I'll have to mull this over.

Reply:
@JamesAPetts key collision is a good call out. If we limit ourselves to one active tool per `interaction paradigm` + `mask`, then having `modifier` keys shouldn't produce collisions. Only passive actions with duplicate bindings would be an issue, right?